tutorials/nservicebus-101/lesson-2/tutorial.md

```diff
@@ -0,0 +1,182 @@
+
+Also, you should not embed logic within your NServiceBus message classes. Each message should contain only automatic properties and not computed properties or methods. Also, it is a good practice to instantiate collection properties from a default parameterless constructor as shown above, so that you never have to deal with a potentially null collection.
+
+In essence, messages should be carriers for data only. This way, discovering a bug in logic will not require redeployment of every endpoint that uses a message.
+
+
+## How do I organize messages?
+
+Because messages must be shared between multiple endpoints, you generally should not put the classes in the same assembly with the endpoints. Instead, they should live in a separate class library so that they can be referenced by multiple endpoints without forcing those endpoints to take in additional dependencies.
+
+A **message assembly** should contain only NServiceBus message contracts, and any supporting types required by the messages themselves. For example, a message may use an enumeration type for one of its properties; that enumeration should also live within the same message assembly.
+
+Message assemblies should be entirely self-contained. They should have no dependencies other than libraries included with the .NET Framework, and the NServiceBus core assembly, which is required to reference the `ICommand` interface. Limiting dependencies makes your message contracts more resilient to future changes in the system.
+
+NOTE: It's also possible to use [message conventions](/nservicebus/messaging/conventions.md) to identify message types, rather than the `ICommand` interface. Then a message assembly doesn't even need to reference NServiceBus at all. This can be an advantage when you want to upgrade to a new major version of NServiceBus, because message assemblies (which are shared between multiple endpoints) don't require any updates, so you can more easily update only one endpoint at a time.
+
+
+## How do I process a message?
+
+To process a message, we create a [**message handler**](/nservicebus/handlers/), a class that implements `IHandleMessages<T>`, where `T` is a message type. A message handler looks like this:
+
+snippet:EmptyHandler
+
+The implementation of the `IHandleMessages<T>` interface is the `Handle` method, which NServiceBus will invoke when a matching message (in this case `DoSomething`) arrives. The `Handle` method receives the message and an `IMessageHandlerContext` that contains tools for working with messages.
+
+Of course, you can add the `async` keyword to a handler method, and then you do not need to explicitly return a task, as shown below. (If you are not comfortable working with async methods, you may want to check out [Asynchronous Handlers](/nservicebus/handlers/async-handlers.md) for more information.)
```

Comment:
> Of course, you can add the `async` keyword to a handler method, and then you do not need to explicitly return a task, as shown below. (If you are not comfortable working with async methods, you may want to check out [Asynchronous Handlers](/nservicebus/handlers/async-handlers.md) for more information.)

To make reader feel more comfortable I'd say:

"Instead of explicitly returning a task, you can add the `async` keyword to a handler method:

snippet:EmptyHandlerAsync

If you want to learn more about working with async methods, check out the [Asynchronous Handlers](/nservicebus/handlers/async-handlers.md) article.